posix + tests: correct timing functions to use CLOCK_REALTIME

cfriedt opened the pull request:
Use `CLOCK_REALTIME` by default as per the POSIX specification for the following functions:

* `mq_timedreceive()`
* `mq_timedsend()`
* `pthread_cond_timedwait()`
* `pthread_mutex_timedlock()`
* `pthread_rwlock_timedrdlock()`
* `pthread_rwlock_timedwrlock()`
* `sem_timedwait()`

and additionally for the following non-portable functions:

* `pthread_timedjoin_np()`

Fixes #88587

Should be merged after #88767 and #88769

cfriedt commented:
An aside, I think it would be really useful to have an API call to perform common operations on `timespec` objects, sort of like https://github.com/solemnwarning/timespec, that additionally reports when an overflow has occurred.

That API and function signatures are good, but I would almost prefer if the entire implementation were in a header file, as static inlines, and additionally be able to use `__builtin_add_overflow()` and friends, as long as C11 was available.

https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html

ycsin reviewed (commented):
Looks like there's some cleanup + code refactoring in the same commit, would it be possible to split this into smaller units?

cfriedt commented:
> Looks like there's some cleanup + code refactoring in the same commit, would it be possible to split this into smaller units? 

It's not easy. There are a lot of dependencies. Little to do with cleanup, but all of it is required to get CLOCK_REALTIME as the default.

jukkar commented on lib/posix/options/posix_clock.h:
```diff
@@ -1,19 +1,82 @@
+extern "C" {
+#endif
+
+static inline bool is_timespec_valid(const struct timespec *ts)
```
Perhaps mark these with functions/macros below that are currently missing doxygen comments by `/** @cond INTERNAL_HIDDEN */`

cfriedt replied:
Fixed!

Commit:
posix: posix_internal: add definitions for posix_condattr, posix_cond

Add definitions for struct posix_condattr and struct posix_cond, which
are internal variants of the external pthread_condattr_t and
pthread_cond_t types.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: cond: use struct posix_cond and struct posix_condattr internally

Use struct posix_cond and struct posix_condattr internally.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: declare timespec_to_timeoutms() in posix_clock.h

Provide a single declaration of timespec_to_timeoutms() (which is
a private function), in the private header file posix_clock.h .

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: cond: check whether condattr is initialized in attr fns

Check whenther a pthread_condattr_t has been initialized in
pthread_condattr_*() functions.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: provide timespec_to_clock_timeoutms

Provide a private internal function timespec_to_clock_timeoutms() to
complement timespec_to_timeoutms(). This new variant accepts a clock_t
parameter that allows the caller to specify which clock to use.

The original timespec_to_timeoutms() then just becomes a static inline
wrapper around the original.

Note: timespec_to_clock_timeoutms() and timespec_to_timeoutms() might
have a limited lifespan, since it might make sense to create a
common timespec manipulation library.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: cond: use clock specified via pthread_condattr_t

Use the clock specified via pthread_condattr_t in
pthread_cond_timedwait().

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: add timespec_is_valid() private internal function

Add a common private function timespec_is_valid() that
can be used to check if a timespec object is valid, and
use that consistently in lib/posix/options.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: pthread: correct cases for pthread_timedjoin_np()

pthread_timedjoin_np() is allowed to take a negative tv_sec
parameter, since that is still a valid timespec.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix + tests: use CLOCK_REALTIME where specified by POSIX

Use CLOCK_REALTIME for the default clock source throughout
the POSIX implementation and tests so that we are
consistent with the specification.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: always require clockid_t argument to timespec_to_timeoutms()

Always require the clockid_t argument to timespec_to_timeoutms() and
remove the unused variant that accepts no clockid_t parameter.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: doc: hide internal posix_clock.h functions from doxygen

Prevent doxygen from parsing internal functions declared in
posix_clock.h .

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: move timespec functions to posix_clock.h

Move somewhat useful (but private and internal functions) that deal
with struct timespec to posix_clock.h until there is a better API
available for dealing with operations on struct timespec.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: refactor timespec_to_timeoutms() to use tp_diff()

Use the tp_diff() macro as a means of converting an absolute timeout
with respect to a specific clock to a relative timeout, in ms.

Clamp the result between 0 and UINT32_MAX.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
* removed clever macro-based timespec functions due to compliance warnings 🫠

cfriedt commented on tests/posix/timers/src/clock.c:
```diff
@@ -29,35 +33,6 @@ static const bool settable[] = {
-		return _tp_op(_a, _b, _op);                                                        \
-	}
-
-_decl_op(bool, tp_eq, ==);     /* a == b */
-_decl_op(bool, tp_lt, <);      /* a < b */
-_decl_op(bool, tp_gt, >);      /* a > b */
-_decl_op(bool, tp_le, <=);     /* a <= b */
-_decl_op(bool, tp_ge, >=);     /* a >= b */
-_decl_op(int64_t, tp_diff, -); /* a - b */
```
Had to remove these to avoid
```
MACRO_WITH_FLOW_CONTROL: Macros with flow control statements should be avoided
```

ycsin replied:
some of these aren't in the `posix_clock.h` header file, causing the compilation error in CI

cfriedt replied:
Fixed!

cfriedt commented:
* replaced `tp_ge()` in `posix_clock.h`

kartben commented on lib/posix/options/cond.c:
```diff
@@ -163,27 +170,39 @@ int pthread_cond_broadcast(pthread_cond_t *cvar)
 {
-	struct k_condvar *cv;
+	struct posix_cond *cv;
+	struct posix_condattr *attr = (struct posix_condattr *)attr;
```
#89076 :)
